plugin/checkout_handlers.go

```diff
@@ -52,6 +53,76 @@ type CheckOutHandler interface {
 	Delete(ctx context.Context, storage logical.Storage, serviceAccountName string) error
 }
 
+// NewServiceAccountLocker is the preferable way to instantiate a ServiceAccountLocker
+// because it populates the map of locks for you.
+func NewServiceAccountLocker(child CheckOutHandler) *ServiceAccountLocker {
+	return &ServiceAccountLocker{
+		locks: &sync.Map{},
+		child: child,
+	}
+}
+
+// ServiceAccountLocker protects against races.
+type ServiceAccountLocker struct {
+	// This is, in effect, being used as a map[string]*sync.RWMutex
+	locks *sync.Map
+	child CheckOutHandler
+}
+
+// CheckOut holds a write lock for the duration of the work to be done.
+func (l *ServiceAccountLocker) CheckOut(ctx context.Context, storage logical.Storage, serviceAccountName string, checkOut *CheckOut) error {
+	if err := validateInputs(ctx, storage, serviceAccountName, checkOut, true); err != nil {
+		return err
+	}
+	lock := l.getOrCreateLock(serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+	return l.child.CheckOut(ctx, storage, serviceAccountName, checkOut)
+}
+
+// CheckIn holds a write lock for the duration of the work to be done.
+func (l *ServiceAccountLocker) CheckIn(ctx context.Context, storage logical.Storage, serviceAccountName string) error {
+	if err := validateInputs(ctx, storage, serviceAccountName, nil, false); err != nil {
+		return err
+	}
+	lock := l.getOrCreateLock(serviceAccountName)
+	lock.Lock()
+	defer lock.Unlock()
+	return l.child.CheckIn(ctx, storage, serviceAccountName)
+}
+
+// Delete holds a write lock for the duration of the work to be done.
+func (l *ServiceAccountLocker) Delete(ctx context.Context, storage logical.Storage, serviceAccountName string) error {
```

Comment:
- `ServiceAccountLocker.Delete` calls `validateInputs()`, then calls `l.child.Delete`. 
- If the child is a PasswordHandler:
  - `PasswordHandler.Delete` calls `validateInputs()`, then calls `h.child.Delete`
  - assuming `PasswordHandler`'s child is a `StorageHandler` (as it's the only other type at the moment(?)):
    - `StorageHandler.Delete` calls `validateInputs()`
- If the child is a `StorageHandler`:
  - `StorageHandler.Delete` calls `validateInputs()`

I feel like we may be validating inputs too much. At best we call it twice.

Reply:
Ah, I misunderstood, I thought you wanted me to add it on every exported method. I agree, it's quite repetitive. Since it's a discrete task, to follow the pattern of the stack where each discrete task is its own thing, I will move it out and set it at the front. It will increase clarity, reduce repetition, and follow the model.

Reply:
This is addressed in https://github.com/hashicorp/vault-plugin-secrets-ad/pull/44/commits/b2edb48d31490fe0918337c5ede84280290044c3, and I'll add it as the first thing in the stack of handlers.